### PR TITLE
[READY] Ports chem dispenser macro recording

### PR DESCRIPTION
**Pull request by Arturlang**
Ports https://github.com/tgstation/tgstation/pull/45397, which replaces macro typing with recording, which is more simpler to use and user friendly, while making powergaming a little bit harder.
## Changelog
:cl:
del: Removed text macros from the chem dispenser.
add: Replaced with dispenser input recording macros.
/:cl:

### CODE REVIEW

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 1]
```diff
@@ -87,20 +87,17 @@
...
 	var/list/saved_recipes = list()
 
 /obj/machinery/chem_dispenser/Initialize()
 	. = ..()
+	dispensable_reagents = sortList(dispensable_reagents, /proc/cmp_reagents_asc)
 	if(emagged_reagents)
 		emagged_reagents = sortList(emagged_reagents, /proc/cmp_reagents_asc)
 	if(upgrade_reagents)
 		upgrade_reagents = sortList(upgrade_reagents, /proc/cmp_reagents_asc)
-	if(upgrade_reagents2)
```
Comment (Putnam3145) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: You sure about that? From looking elsewhere in the code, I think this means that you get NO upgraded reagents until you've FULLY upgraded your chem dispenser, which is kinda weird.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 2]
```diff
@@ -226,7 +213,6 @@
 		data["beakerCurrentpH"] = null
 
 	var/chemicals[0]
```
Comment (Putnam3145) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: While you're at it: `var/list/chemicals = list()`

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 3]
```diff
@@ -422,41 +412,13 @@
...
-		final_list += list(avoid_assoc_duplicate_keys(fuck[1],key_list) = text2num(fuck[2]))
-	return final_list
-
 /obj/machinery/chem_dispenser/AltClick(mob/living/user)
-	. = ..()
-	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
-		return
-	replace_beaker(user)
-	return TRUE
+	..()
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]:
```suggestion
	. = ..()
```

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 4]
```diff
@@ -422,41 +412,13 @@
...
+	if(istype(user) && user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+		replace_beaker(user)
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm:430]: add a `return TRUE` after this line, same indentation.

**Review by Ghommie — CHANGES_REQUESTED** [review]
The macro tier amount precision removal makes sense since it records pressed buttons now, and it's not really possible to go below 5u transfers with a normal beaker (plant nutriment bottles anyone?), but as Putnam pointed out, `upgrade_reagents2` and `upgrade_reagents3` should stay around.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm (hunk 1 above)
Comment (Arturlang) [reply code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Whoops, seems i fucked that up.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 5]
```diff
@@ -395,14 +396,8 @@
...
 	for(var/obj/item/stock_parts/manipulator/M in component_parts)
-		if (M.rating > macrotier)
-			macrotier = M.rating
-		if (M.rating > 1)
+		if(M.rating > 3)
 			dispensable_reagents |= upgrade_reagents
-		if (M.rating > 2)
-			dispensable_reagents |= upgrade_reagents2
-		if (M.rating > 3)
-			dispensable_reagents |= upgrade_reagents3
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: You forgot about this bit.

**Comment by Hatterhat** [comment]
fuck med mains

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 6]
```diff
@@ -396,8 +396,14 @@
 	for(var/obj/item/stock_parts/capacitor/C in component_parts)
 		recharge_amount *= C.rating
 	for(var/obj/item/stock_parts/manipulator/M in component_parts)
-		if(M.rating > 3)
+		if(M.rating > macrotier)
+			macrotier = M.rating
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: Just remove.

File: code/modules/reagents/chemistry/machinery/chem_dispenser.dm [hunk 7]
```diff
@@ -31,8 +31,9 @@
 	var/working_state = "dispenser_working"
 	var/nopower_state = "dispenser_nopower"
 	var/has_panel_overlay = TRUE
-	var/macrotier = 1
 	var/obj/item/reagent_containers/beaker = null
+	var/macrotier = 1
```
Comment (Ghommie) [inline code/modules/reagents/chemistry/machinery/chem_dispenser.dm]: remove this.

**Review by Ghommie — CHANGES_REQUESTED** [review]
`macrotier` is useless now.

**Review by Ghommie — APPROVED** [review]

**Comment by InquisitiveEpidemic** [comment]
can you please fix that it is constantly moving the UI as it's being used. You can't reliably hit the same chem twice with this. That said, I fully disagree with cutting macro because we are removing levels of precision in reagent dispensing that WAS ALREADY ACCOUNTED FOR WITH TECH TIERS, and it will do nothing to actually affect powergame.

**Pull request merged**